File: frontend/iosApp/UseCases/Sources/UseCases/Impls/AlbumDetailUseCase.swift
import Foundation
import Combine
import Domains
import Repositories
import APIGateways

public final class AlbumDetailUseCase: AlbumDetailUseCaseProtocol, @unchecked Sendable {
    private let memoryRepository: MemoryRepositoryProtocol
    private let albumRepository: AlbumRepositoryProtocol
    private let memoryGateway: MemoryGatewayProtocol
    private let reachabilityRepository: ReachabilityRepositoryProtocol

    public var localChangePublisher: AnyPublisher<LocalMemoryChangeEvent, Never> {
        memoryRepository.localChangePublisher
    }

    public var observeAlbumUpdate: AnyPublisher<LocalAlbumChangeEvent, Never> {
        albumRepository.localChangePublisher
    }

    public init(
        memoryRepository: MemoryRepositoryProtocol,
        albumRepository: AlbumRepositoryProtocol,
        memoryGateway: MemoryGatewayProtocol,
        reachabilityRepository: ReachabilityRepositoryProtocol
    ) {
        self.memoryRepository = memoryRepository
        self.albumRepository = albumRepository
        self.memoryGateway = memoryGateway
        self.reachabilityRepository = reachabilityRepository
    }

    public func display(album: Album) async -> AlbumDetailUseCaseModel.DisplayResult {
        // If album is not synced yet, only show local memories
        guard let albumServerId = album.id else {
            let cached = await memoryRepository.getAll(albumLocalId: album.localId)
            return .success(AlbumDetailUseCaseModel.PageInfo(memories: cached, hasMore: false))
        }

        if reachabilityRepository.isConnected {
            do {
                let response = try await memoryGateway.getMemories(albumId: albumServerId, page: 1, pageSize: Const.pageSize)
                let memories = response.items.compactMap { MemoryMapper.toDomain($0, albumLocalId: album.localId) }
                try? await memoryRepository.syncSet(memories, albumLocalId: album.localId)
                let allMemories = await memoryRepository.getAll(albumLocalId: album.localId)
                let hasMore = response.page * response.pageSize < response.total
                return .success(AlbumDetailUseCaseModel.PageInfo(memories: allMemories, hasMore: hasMore))
            } catch {
                // Fallback to cache on error
                let cached = await memoryRepository.getAll(albumLocalId: album.localId)
                if !cached.isEmpty {
                    return .success(AlbumDetailUseCaseModel.PageInfo(memories: cached, hasMore: false))
                }
                return .failure(mapDisplayError(error))
            }
        } else {
            // Offline: get from cache (empty is valid - album may have no memories yet)
            let cached = await memoryRepository.getAll(albumLocalId: album.localId)
            return .success(AlbumDetailUseCaseModel.PageInfo(memories: cached, hasMore: false))
        }
    }

    public func next(album: Album, page: Int) async -> AlbumDetailUseCaseModel.NextResult {
        // Pagination requires online and synced album
        guard reachabilityRepository.isConnected else {
            return .failure(.offline)
        }

        guard let albumServerId = album.id else {
            return .failure(.offline)
        }

        do {
            let response = try await memoryGateway.getMemories(albumId: albumServerId, page: page, pageSize: Const.pageSize)
            let memories = response.items.compactMap { MemoryMapper.toDomain($0, albumLocalId: album.localId) }
            try? await memoryRepository.syncAppend(memories)
            let allMemories = await memoryRepository.getAll(albumLocalId: album.localId)
            let hasMore = response.page * response.pageSize < response.total
            return .success(AlbumDetailUseCaseModel.PageInfo(memories: allMemories, hasMore: hasMore))
        } catch {
            return .failure(mapNextError(error))
        }
    }

    private func mapDisplayError(_ error: Error) -> AlbumDetailUseCaseModel.DisplayResult.Error {
        let nsError = error as NSError
        if nsError.domain == NSURLErrorDomain {
            return .networkError
        }
        return .unknown
    }

    private func mapNextError(_ error: Error) -> AlbumDetailUseCaseModel.NextResult.Error {
        let nsError = error as NSError
        if nsError.domain == NSURLErrorDomain {
            return .networkError
        }
        return .unknown
    }
}

// MARK: - Const

extension AlbumDetailUseCase {
    private enum Const {
        static let pageSize = 5
    }
}

File: frontend/shared/src/commonMain/kotlin/com/example/memoriesapp/usecase/AlbumDetailUseCase.kt
package com.example.memoriesapp.usecase

import com.example.memoriesapp.domain.Album
import com.example.memoriesapp.domain.Memory
import com.example.memoriesapp.repository.LocalAlbumChangeEvent
import com.example.memoriesapp.repository.LocalMemoryChangeEvent
import kotlinx.coroutines.flow.Flow

/**
 * Memory page info
 */
data class MemoryPageInfo(
    val memories: List<Memory>,
    val hasMore: Boolean
)

/**
 * Result of display memories operation
 */
sealed class MemoryDisplayResult {
    data class Success(val pageInfo: MemoryPageInfo) : MemoryDisplayResult()
    data class Failure(val error: MemoryDisplayError) : MemoryDisplayResult()
}

enum class MemoryDisplayError {
    OFFLINE,
    NETWORK_ERROR,
    UNKNOWN
}

/**
 * Result of next page operation for memories
 */
sealed class MemoryNextResult {
    data class Success(val pageInfo: MemoryPageInfo) : MemoryNextResult()
    data class Failure(val error: MemoryNextError) : MemoryNextResult()
}

enum class MemoryNextError {
    OFFLINE,
    NETWORK_ERROR,
    UNKNOWN
}

/**
 * Result of resolving an album by server ID
 */
sealed class ResolveAlbumResult {
    data class Success(val album: Album) : ResolveAlbumResult()
    data class Failure(val error: ResolveAlbumError) : ResolveAlbumResult()
}

enum class ResolveAlbumError {
    NOT_FOUND,
    NETWORK_ERROR,
    OFFLINE_UNAVAILABLE
}

interface AlbumDetailUseCase {
    val localChangeFlow: Flow<LocalMemoryChangeEvent>
    val observeAlbumUpdate: Flow<LocalAlbumChangeEvent>
    suspend fun display(album: Album): MemoryDisplayResult
    suspend fun next(album: Album, page: Int): MemoryNextResult
    suspend fun resolveAlbum(serverId: Int): ResolveAlbumResult
}
